### ISSUE TITLE
Failure on OpenBSD 7.3 because read-only data is in .text instead of .rodata

### ISSUE THREAD

**Issue by siegfried** [briansmith/ring#1602]
I'm getting segment fault after upgrading OpenBSD from 7.2 to 7.3. Here is an error I found in core dump:

Reading symbols from /usr/libexec/ld.so...Error while reading shared library symbols:
Dwarf Error: wrong version in compilation unit header (is 4, should be 2) [in module /usr/libexec/ld.so]
#0  GFp_sha512_block_data_order_avx () at /mnt/warehouse/home/git/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/sha512-x86_64-elf.S:1881
1881		vpaddq	-128(%rbp),%xmm0,%xmm8

**Comment by kylegodbold** [comment]
Use rust-gdb instead of the normal gdb, then you won't get the Dwarf Error when opening up the core file.

I'm getting a very similar error on the new OpenBSD 7.3 release. The same code compiles and runs just fine on OpenBSD 7.2 but that version uses Rust 1.63 while 7.3 has the package for Rust 1.68. I'm not sure if it's the Rust compiler or OpenBSD but I'm trying to figure out what change between then and now introduced the error. Ring 0.16.20 works fine outside of 7.3 but just like every OpenBSD release there's new exploit mitigations which could be what's triggering it. I'll post here when I have something more concrete.

This is the error that I'm getting:

```Program terminated with signal SIGSEGV, Segmentation fault.
#0  GFp_sha256_block_data_order_ssse3 ()
    at /home/kyle/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/sha256-x86_64-elf.S:1824
1824            movdqa  K256+512(%rip),%xmm7
[Current thread is 1 (process 405501)]
(gdb) bt
#0  GFp_sha256_block_data_order_ssse3 ()
    at /home/kyle/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/sha256-x86_64-elf.S:1824
#1  0x00000facc5428527 in ring::digest::Context::update ()
#2  0x00000facc5457f1f in rustls::hash_hs::HandshakeHash::add_message ()
#3  0x00000facc5462ae1 in rustls::server::hs::ExpectClientHello::with_certified_key ()
#4  0x00000facc545ed04 in <rustls::server::hs::ExpectClientHello as rustls::conn::State<rustls::server::server_conn::ServerConnectionData>>::handle ()
#5  0x00000facc54ecb16 in rustls::conn::CommonState::process_main_protocol ()
#6  0x00000facc54eadf8 in tokio_rustls::common::Stream<IO,C>::read_io ()
#7  0x00000facc54e6ca1 in <tokio_rustls::Accept<IO> as core::future::future::Future>::poll ()
#8  0x00000facc54e6980 in <warp::tls::TlsStream as tokio::io::async_read::AsyncRead>::poll_read ()
#9  0x00000facc51c2a13 in hyper::proto::h1::io::Buffered<T,B>::poll_read_from_io ()
#10 0x00000facc51d3e5c in hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_loop ()
#11 0x00000facc5158a28 in <hyper::server::server::new_svc::NewSvcTask<I,N,S,E,W> as core::future::future::Future>::poll ()
#12 0x00000facc51e7407 in tokio::runtime::task::raw::poll ()
#13 0x00000facc54bc3c9 in tokio::runtime::scheduler::multi_thread::worker::Context::run_task ()
#14 0x00000facc54c31c3 in tokio::runtime::task::raw::poll ()
#15 0x00000facc54b4359 in std::sys_common::backtrace::__rust_begin_short_backtrace ()
#16 0x00000facc54b40fd in core::ops::function::FnOnce::call_once{{vtable-shim}} ()
#17 0x00000facc54812f9 in std::sys::unix::thread::Thread::new::thread_start ()
#18 0x00000fafa0b7ce51 in _rthread_start (v=<optimized out>)
    at /usr/src/lib/librthread/rthread.c:96
#19 0x00000faed54b8d6a in __tfork_thread ()
    at /usr/src/lib/libc/arch/amd64/sys/tfork_thread.S:86
```

**Comment by kylegodbold** [comment]
According to [Theo Buehler on the OpenBSD-misc mailing list](https://marc.info/?l=openbsd-misc&m=168115293527678&w=2):

```
Another issue could be that you use things depending on the ring crate
whose assembly isn't compatible with the new x-only protection on modern
amd64 machines.

If so, you will need to 

# pkg_add rust-ring

and replace lines like this

ring         = "^0.16"

with something like this in Cargo.toml:

[dependencies.ring]
version         = "^0.16"
path = "/usr/local/share/ring-0.16.20"
```

It's nice to know that it's yet another security new security feature, even though it's been a bit of a pain for me the past few days since I've upgraded. No pain no gain.

**Comment by n1000** [comment]
For reference, [here are the patches](https://github.com/openbsd/ports/tree/master/security/rust-ring/patches
) applied in the OpenBSD rust-ring package.

It looks like a majority of the patches just mark some of the data used in the pregenerated assembly files as being in the .rodata section. A couple of the patches such as `patch-pregenerated_ecp_nistz256-armv8-linux64_S` and `patch-crypto_fipsmodule_ec_asm_ecp_nistz256-armv8_pl` also modify some of the actual instructions used slightly...

**Comment by n1000** [comment]
In addition to what @kylegodbold mentioned, another email on the openbsd misc mailing list suggested adding the packaged rust-ring dependency to the top level `~/.cargo/config.toml` so that local builds can automatically pick up the patched version. Excerpt from [that email](https://marc.info/?l=openbsd-misc&m=168120112917160&w=2):

```
In case it's useful to anyone else, one can set this globally in
~/.cargo/config.toml (and avoid tweaking multiple Cargo.toml's, which can be
a pain with dependencies) with:

  [patch.crates-io]
  ring = { path = "/usr/local/share/ring-0.16.20" }

However, there is one gotcha: this will cause your Cargo.lock to be altered,
so if you're working on a project with a checked-in Cargo.lock, you need to
be careful.
```

Just adding this here, since this seems to be the most useful short term solution to keep most builds working.

**Comment by 22h49** [comment]
Having the same issue on OpenBSD 7.3, in the same asm file. +1

**Comment by niklasha** [comment]
Just wanted to add that I also tripped on this last night.  It is obviously correct to move constant data from .text to .rodata, as perhaps more operating systems will go the way of make .text execute-only (i.e not allowing reads) in the future.  Even before then, the data can not be sought for coincidental pieces of raw bytes, that happen to be valid instructions, doing something nasty, reachable by an exploit changing the return pointer on the stack just before a return instruction.

**Comment by botovq** [comment]
I would be happy to create a PR with these patches. I just did not yet have time to look into it.

Similar patches have been merged in BoringSSL: [1][1] [2][2]. The better option may be to sync the perlasm with BoringSSL to avoid further divergence.

[1]: https://github.com/google/boringssl/commit/ebd43ef8dd3c062dccf6adf264c5332efda1f4b3
[2]: https://github.com/google/boringssl/commit/e18ba272d4532659a20904c812207079c4ec2e80

**Comment by briansmith** [comment]
> The better option may be to sync the perlasm with BoringSSL to avoid further divergence.

I agree that that is the best course of action. Thanks for reporting this and thanks for investigating it.

**Comment by labannah9125** [comment]
why openbsd not provide patches upstream here if they make sense? i run into this today and am hard blocked on my app at the moment due to needing something with rustls using ring.

**Comment by niklasha** [comment]
hard blocked?  Are you on OpenBSD?  Then a run of `pkg_add rust-ring` and the addition of
```
[patch.crates-io]
ring = { path = "/usr/local/share/ring-0.16.20" }
```
in your `~/.cargo/config` will help, until ring syncs with BoringSSL's perlasm, which I thought was the plan...  Of course anyone can take OpenBSD's rust-ring package's patches and make a PR here.  I'd do it if I didn't think the resolution plan was already fixed.

**Comment by labannah9125** [comment]
> Of course anyone can take OpenBSD's rust-ring package's patches and make a PR here

original patcher probably should have when making for openbsd anyway as is written in their port policy guide. couldnt hurt doing both? maybe save the packager work keeping them updated later even if they did it themselves here.

**Comment by niklasha** [comment]
I won't argue about that it would've been nice. I just wanted to question the "hard blocked" statement.

On a note about the complaint:  It is at times not very gratifying work to do open source development that people may use at will, and then be told they fail perfection. I am happy someone went ahead and did the work before I needed to, and make it available _somewhere_.  Of course it would have been nice if it was to be found the first place we look, but it's not like we can demand it. Open source is about sharing work for others to build upon, it's not about signing into a support contract requiring people to do things they have no time/gratification for.

**Comment by briansmith** [comment]
> > The better option may be to sync the perlasm with BoringSSL to avoid further divergence.
> I agree that that is the best course of action. Thanks for reporting this and thanks for investigating it.

*ring* 0.17.0 has synced BoringSSL up to the latest commit, and incorporated these changes.

**Comment by briansmith** [comment]
> will someone please still bring in openbsd patch to avoid future issue?

We are working to upgrade everything to 0.17.0.

**Comment by labannah9125** [comment]
> > will someone please still bring in openbsd patch to avoid future issue?
> 
> We are working to upgrade everything to 0.17.0.

with linked patches from openbsd repo?

**Comment by briansmith** [comment]
According to @botovq, similar patches were merged into BoringSSL. I merged BoringSSL into *ring* 0.17 so I would expect them to be in *ring*. It would be great if somebody would double-check AND report here for others.

**Comment by botovq** [comment]
> It would be great if somebody would double-check AND report here for others.

The patches are present and unpatched ring 0.17 passes regress on
OpenBSD amd64 and aarch64. amd64 was also tested on an Intel machine
with IBT.

**Comment by briansmith** [comment]
> The patches are present and unpatched ring 0.17 passes regress on OpenBSD amd64 and aarch64. amd64 was also tested on an Intel machine with IBT.

Awesome! Which CPU and OS version were you using?

**Comment by botovq** [comment]
> Awesome! Which CPU and OS version were you using?

The tests I did myself were on OpenBSD 7.4-current self-built from HEAD
source in the last few days.

aarch64 was tested on an M1 mac mini from 2021 (too old to support the
bti instruction):

```
cpu0 at mainbus0 mpidr 0: Apple Icestorm r1p1
```

and amd64 was tested on

```
cpu0: Intel(R) Core(TM) i7-10700K CPU @ 3.80GHz, 3691.40 MHz, 06-a5-05, patch 000000f8
```

and

```
cpu0: AMD Ryzen 7 PRO 5850U with Radeon Graphics, 1900.00 MHz, 19-50-00, patch 0a50000d
```

Both these CPUs support PKU, so exercise the .rodata paths this issue
is about, but they don't exercise the CET bits.

The tests on an IBT-enabled Intel machine were performed by @ltratt. I
don't know specifics, but it will have been some flavor of -current.

**Comment by ltratt** [comment]
> The tests on an IBT-enabled Intel machine were performed by [@ltratt](https://github.com/ltratt). I
don't know specifics, but it will have been some flavor of -current.

Yes, I tested on an Intel 13600K, with current amd64 as of about a week ago.

**Comment by briansmith** [comment]
Thanks for reporting back. Glad this is working now!

**Comment by botovq** [comment]
Thanks for doing the heavy lifting and fixing this as a side-effect.

**Comment by brahin2** [comment]
i had same issue. does it mean no-longer needing the system package and that https://github.com/openbsd/ports/tree/master/security/rust-ring/patches were merged in here or boringssl? or will there still be issue when not using the hacked up package?
